exercises/smartPointers/problem3.cpp

```diff
@@ -0,0 +1,136 @@
+
+
+#include <iostream>
+#include <array>
+#include <vector>
+#include <random>
+#include <algorithm>
+
+
+/* --------------------------------------------------------------------------------------------
+ * Shared ownership.
+ *
+ * Most of the time, ownership can be solved by having one owner (with `std::unique_ptr`) and
+ * one or more observers (raw pointers or references). Sometimes, we need to truly share data,
+ * though. Here is an example of a completely messed up ownership model.
+ *
+ * Tasks
+ * 1) Verify the mess by repeatedly running it using such a command like:
+ *   `while true; do ./problem3 2>&1 ; done`
+ *    You should notice that the program regularly leaks.
+ * 2) Fix the ownership model using `std::shared_ptr` !
+ *   - Convert the vectors to holding `std::shared_ptr`.
+ *   - Fix the arguments of the functions.
+ * 3) Speed optimisation: make sure that you don't create & destroy useless `std::shared_ptr`,
+ *    which is slow, for example in the for loop of `doStuff()` and when
+ *    calling `changeLargeObject()`.
+ * --------------------------------------------------------------------------------------------
+ */
+
+
+// The class LargeObject emulates a large object.
+// One should avoid to copy it, and rather use
+// a pointer to pass it around.
+
+struct LargeObject {
+
+    std::array<double, 100000> data ;
+
+    // So to check for some potential memory leak,
+    // we count the constructions and destructions
+    static std::size_t count ;
+    LargeObject() { count++ ; }
+    ~LargeObject() { count-- ; }
+
+} ;
+
+std::size_t LargeObject::count = 0 ;
+
+// This removes an element from a non-owning vector,
+// in the middle or in a random place. Such elements
+// can by known in several vectors, so they must not
+// be deleted.
+
+void removeMiddle( std::vector<LargeObject *> & collection ) {
+
+    // MAKE YOUR CHANGES IN THIS FUNCTION
+
+    auto middlePosition = collection.begin() + collection.size()/2 ;
+    collection.erase(middlePosition) ;
+
+}
```

Comment:
Could we remove this maybe?

I think it has only historic value, but it doesn't add anything to the exercise.

Reply:
True.